standard/io/witx/io_streams.witx

```diff
@@ -1,32 +1,256 @@
 ;; WASI I/O Streams.
+;;
+;; I/O streams support reliable in-order delivery.
+;;
+;; Currently this interface only supports streams of bytes, and datagram
+;; sockets and out-of-band messages are not supported. In the future, with
+;; appropriate interface types support, it's expected that this interface
+;; will support streams of arbitrary type as well, which could be implemented
+;; on top of datagram sockets.
+;;
+;; Many parts of this file are expected to evolve, as witx evolves, as
+;; interface types evolves, and as the I/O Types concepts themselves evolve.
+;; For example, `iovec_array` will likely be replaced by something else. The
+;; linear-memory import will go away. typenames.witx will likely be obviated
+;; with better mechanisms for sharing types.
 
 (use "typenames.witx")
 
+;;; A handle providing reliable and in-order delivery of a stream of bytes
+;;; from an external source.
+(typename $input_byte_stream (handle))
+
+;;; A handle providing reliable and in-order delivery of a stream of bytes
+;;; to an external source.
+(typename $output_byte_stream (handle))
+
+;;; A handle representing the name of a stream. This is opaque, and does
+;;; not expose the actual string.
+(typename $pseudonym (handle))
+
+;;; The status of a stream after a read.
+(typename $read_status
+  (enum (@witx tag u8)
+    ;;; The stream is ready for further reads.
+    ;;;
+    ;;; This code also implies that the `$size` returned from a read is equal
+    ;;; to the requested size.
+    $ready
+
+    ;;; Push - The stream is ready for further reads, however the source has
+    ;;; performed a flush. Consumers are encouraged to process the available
+    ;;; data before attempting further reads.
+    ;;;
+    ;;; This may also indicate that a failure has been detected after some
+    ;;; amount of data has been consumed. The data is returned successfully
+    ;;; with a `$push`, and the next `read` or `skip` call will report the
+    ;;; failure.
+    $push
```

Comment:
Should there be separate statuses here for streams that model the `PUSH` flag (like TCP) and a different one for indicating a possible failure?
What is the thinking behind having the same status for both?

Reply:
These are good questions :-). I filed https://github.com/WebAssembly/wasi-io/issues/9 to think about these further.